python-frontend/src/test/java/org/sonar/python/semantic/MatchStatementSymbolsTest.java

```diff
@@ -0,0 +1,177 @@
+
+public class MatchStatementSymbolsTest {
+
+  @Test
+  public void capture_pattern() {
+    CapturePattern capturePattern = patternFromCase("case others: print(others)");
+    Symbol others = capturePattern.name().symbol();
+    assertThat(others.name()).isEqualTo("others");
+    assertThat(others.fullyQualifiedName()).isNull();
+    assertThat(others.usages()).extracting(Usage::kind).containsExactly(PATTERN_DECLARATION, OTHER);
+  }
+
+  @Test
+  public void wildcard_pattern() {
+    Name wildcard = getLastDescendant(PythonTestUtils.parse(
+      "def foo(value):",
+      "  match(value):",
+      "    case _: print(_)"
+    ), t -> t.is(NAME) && ((Name) t).name().equals("_"));
+    assertThat(wildcard.symbol()).isNull();
+  }
+
+  @Test
+  public void class_pattern() {
+    ClassPattern classPattern = pattern(
+      "class A:",
+      "  foo = 42",
+      "def foo(value):",
+      "  match(value):",
+      "    case A(foo=x):",
+      "      print(x)");
+    Name name = (Name) classPattern.targetClass();
+    assertThat(name.symbol().kind()).isEqualTo(Symbol.Kind.CLASS);
+    KeywordPattern keywordPattern = ((KeywordPattern) classPattern.arguments().get(0));
+    Symbol x = ((CapturePattern) keywordPattern.pattern()).name().symbol();
+    assertThat(x.usages()).extracting(Usage::kind).containsExactly(PATTERN_DECLARATION, OTHER);
+
+    classPattern = pattern(
+      "import mod",
+      "def foo(value):",
+      "  match(value):",
+      "    case mod.A(foo=x):",
+      "      print(x)");
+    QualifiedExpression qualifiedExpression = (QualifiedExpression) classPattern.targetClass();
+    assertThat(qualifiedExpression.symbol().kind()).isEqualTo(Symbol.Kind.OTHER);
+  }
+
+  @Test
+  public void value_pattern() {
+    ValuePattern valuePattern = pattern(
+      "import command",
+      "def foo(value):",
+      "  match(value):",
+      "    case command.QUIT:",
+      "      print(x)");
+    assertThat(valuePattern.qualifiedExpression().symbol().kind()).isEqualTo(Symbol.Kind.OTHER);
+  }
+
+  @Test
+  public void as_pattern() {
+    AsPattern asPattern = patternFromCase("case 42 as x: ...");
+    Symbol symbol = asPattern.alias().name().symbol();
+    assertThat(symbol.name()).isEqualTo("x");
+    assertThat(symbol.fullyQualifiedName()).isNull();
+    assertThat(symbol.usages()).extracting(Usage::kind).containsExactly(PATTERN_DECLARATION);
+
+    asPattern = patternFromCase("case z as x: return x + z");
+    Symbol z = ((CapturePattern) asPattern.pattern()).name().symbol();
+    Symbol x = asPattern.alias().name().symbol();
+    assertThat(z.name()).isEqualTo("z");
+    assertThat(x.name()).isEqualTo("x");
+    assertThat(z.usages()).extracting(Usage::kind).containsExactly(PATTERN_DECLARATION, OTHER);
+    assertThat(x.usages()).extracting(Usage::kind).containsExactly(PATTERN_DECLARATION, OTHER);
+  }
+
+  @Test
+  public void or_pattern() {
+    OrPattern orPattern = patternFromCase("case A(x) | y: print(x + y)");
```

Comment:
I have a syntax error with this construct:
```
SyntaxError: alternative patterns bind different names
```
I assume some similar case could never happen, as one case would always either not match or get short circuited.
Perhaps it would be more realistic to just test non-binding usages for or patterns?